doc/decisions/DR023-Versioning-traits-and-specifications.md

```diff
@@ -0,0 +1,702 @@
+#
+# The manager is responding to a resolve, needing to populate a
+# TraitsData with a camera trait and a spatial trait of the expected
+# version
+cameraVersion = mediacreation.versionFromTraitId(mediacreation.CameraTrait.kId)
+spatialVersion = mediacreation.versionFromTraitId(mediacreation.SpatialTrait.kId)
+
+# Support all known major versions.
+# Camera
+success_result = TraitsData()
+if (cameraVersion >= 3 and cameraVersion < 4):
+     trait = mediacreation.CameraTrait(success_result)
+     ... # Up to date camera population
+elif (cameraVersion >= 2 and cameraVersion < 1):
+     trait = mediacreation.legacy.CameraTrait_v2(success_result)
+     ... # V2 camera population
+elif (cameraVersion >= 1):
+     trait = mediacreation.legacy.CameraTrait_v1(success_result)
+     ... # V1 camera population
+else:
+  raise RuntimeError()
+
+# Spatial
+if (spatialVersion >= 2 and spatialVersion < 3):
+    trait = mediacreation.SpatialTrait(success_result)
+    ... # Up to date spatial population
+elif (spatialVersion >= 1):
+    trait = mediacreation.legacy.SpatialTrait_v1(success_result)
+     ... # V1 to date spatial population
+else:
+  raise RuntimeError()
+```
+
+### Benefits
+- Maximally flexible, hosts and managers can disregard versioning for
+  traits they don't care about.
+- Obvious from the YAML which version a trait is.
+- Individual traits will only update version when they change, thus they
+  will not have "gaps", they will always go v1->v2->v3.
+- A single trait can be updated in isolation without any effect on other
+  traits in the package.
+- Semver could be used to provide additional runtime compatibility
+  information.
+
+### Drawbacks
+- Version has to be encoded in `traitID` because several methods
+  (including `resolve`) take `traitSet`, which is only a collection of
+  `traitIDs`. There is no `TraitsData` to inspect. This would be a major
+  breaking change as currently conceived.
+- Version support may become fragmented if hosts and managers provide
+  versioned support for only certain traits
+- When packing multiple traits, manager has to branch on each trait as
+  they could be different versions.
+- When extended to specification this _really_ causes problems, as we
+  need to invent a new place to pack specifications version across the
+  wire.
+
+## Version the whole schema.
+This option places a version number is placed at the top of the
+`traits.yml` file or equivalent. This defines the version for all the
+traits within the package.
+
+The manager will still need a mechanism to know which trait-library
+version the host will be expecting. This could be achieved by a
+`HostInterface` method.
+
+```cpp
+//HostInterface implementation
+SchemaVersion HostInterface::schemaVersion(std::string_view library_identifier) {
+  if (library_identifier == "openassetio-mediacreation"){
+    return openassetio_mediacreation::version();
+  }
+}
+```
+### Benefits
+- Version dosen't need to be packed into any `traitId`
```

Comment:
Would this only be true for the host?
Does the manager still need to track the versions? I am assuming multiple apps/hosts can be active at a time, they might all have different version combinations?

Reply:
In any given communication across the API boundary there is only one host communicating with one manager.

As an advanced use case, it is possible for a host to make use of multiple managers at runtime (e.g. for different types of content).  But in any given API call there is only one manager and one host.

A manager only needs to consider multiple hosts in that, ideally, they develop their plugin once and it can be loaded into any host application and "just work".  But at runtime, it only has to deal with one host.

A key (educated) assumption is that hosts will use a single version of a trait package, and it is up to manager plugins to adapt based on the version the host uses.  Therefore a single manager plugin developed for use across multiple different host applications (and different versions of the same host application) may well need to branch in its code, based on the trait package version used by the host it is loaded into.

Reply:
That's a really good point, there may well be 3rd party libraries (perhaps even studio-specific plugins) using the host-side API :thinking: